File: Mixins/src/com/company/keliones/Kelione.scala
package com.company.keliones

import java.math.BigDecimal

import com.company.enums.Intervalas
import com.company.kainosSkaiciavimas.KainosSkaiciavimas

abstract class Kelione(val priemoneId: Int, val isvykimoTaskas: String) extends KainosSkaiciavimas with TransportoPriemone {
  var atvykimoTaskas: String = _
  var atstumas = .0
  var laikas = .0
  private var kaina: BigDecimal = _
  private var ilgalaikisIntervalas: Intervalas.Value = _

  /**
   * @return Keliones kaina.
   */
  def uzbaigtiKelione(atstumas: Double, laikas: Double, atvykimoTaskas: String): BigDecimal = {
    if (atvykimoTaskas.isEmpty) throw new UnsupportedOperationException("Nenustatytas atvykimo taskas.")
    var sumineKaina = apskaiciuotiKelionesKaina(atstumas, laikas, gautiIkainius)
    sumineKaina = koreguotiSumineKaina(isvykimoTaskas, atvykimoTaskas, sumineKaina)
    paliktiTransportoPriemone(priemoneId)
    this.kaina = sumineKaina
    this.atstumas = atstumas
    this.laikas = laikas
    this.atvykimoTaskas = atvykimoTaskas
    sumineKaina
  }

  def uzbaigtiNuoma(atstumas: Double, intervalas: Intervalas.Value): BigDecimal = {
    if (intervalas == null) throw new UnsupportedOperationException("Nenustatytas nuomos intervalas.")
    val ikainiai = gautiIkainius
    var sumineKaina = apskaiciuotiKelionesKaina(atstumas, 0, ikainiai)
    sumineKaina = sumineKaina.add(skaiciuotiIlgalaikeKaina(ikainiai, intervalas))
    paliktiTransportoPriemone(priemoneId)
    System.out.println("Baigete transporto priemones nuoma. Nuomos laikas: " + intervalas + ".")
    this.kaina = sumineKaina
    this.ilgalaikisIntervalas = intervalas
    this.atstumas = atstumas
    sumineKaina
  }
}

File: StrategyPattern/src/com/company/Kelione.java
package com.company;

import com.company.enums.Intervalas;
import com.company.kainosSkaiciavimas.Ikainiai;
import com.company.strategijos.KainosSkaiciavimoStrategija;
import com.company.strategijos.TransportoPasirinkimoStrategija;

import java.math.BigDecimal;

public class Kelione {
    private final TransportoPasirinkimoStrategija tps;
    private final KainosSkaiciavimoStrategija kss;

    private final int priemoneId;
    private final String isvykimoTaskas;
    private String atvykimoTaskas;
    private double atstumas;
    private double laikas;
    private BigDecimal kaina;
    private Intervalas ilgalaikisIntervalas;

    public Kelione(int priemoneId, String isvykimoTaskas, TransportoPasirinkimoStrategija tps, KainosSkaiciavimoStrategija kss) throws IllegalArgumentException, UnsupportedOperationException {
        this.tps = tps;
        this.kss = kss;
        this.priemoneId = priemoneId;
        this.isvykimoTaskas = isvykimoTaskas;
    }

    /**
     * @return Keliones kaina.
     */
    public BigDecimal uzbaigtiKelione(double atstumas, double laikas, String atvykimoTaskas) throws UnsupportedOperationException {
        if (atvykimoTaskas.isEmpty()) {
            throw new UnsupportedOperationException("Nenustatytas atvykimo taskas.");
        }
        BigDecimal sumineKaina = kss.apskaiciuotiKelionesKaina(atstumas, laikas, tps.gautiIkainius());
        sumineKaina = kss.koreguotiSumineKaina(isvykimoTaskas, atvykimoTaskas, sumineKaina);
        tps.paliktiTransportoPriemone();
        this.kaina = sumineKaina;
        this.atstumas = atstumas;
        this.laikas = laikas;
        this.atvykimoTaskas = atvykimoTaskas;
        return sumineKaina;
    }

    /**
     * @return Keliones kaina.
     */
    public BigDecimal uzbaigtiNuoma(double atstumas, Intervalas intervalas) throws UnsupportedOperationException {
        if (intervalas == null) {
            throw new UnsupportedOperationException("Nenustatytas nuomos intervalas.");
        }
        Ikainiai ikainiai = tps.gautiIkainius();
        BigDecimal sumineKaina = kss.apskaiciuotiKelionesKaina(atstumas, 0, ikainiai);
        sumineKaina = sumineKaina.add(kss.skaiciuotiIlgalaikeKaina(ikainiai, intervalas));
        tps.paliktiTransportoPriemone();
        System.out.println("Baigete transporto priemones nuoma. Nuomos laikas: " + intervalas + ".");
        this.kaina = sumineKaina;
        this.ilgalaikisIntervalas = intervalas;
        this.atstumas = atstumas;
        return sumineKaina;
    }

    public String getIsvykimoTaskas() {
        return isvykimoTaskas;
    }

    public int getPriemoneId() {
        return priemoneId;
    }

    public String gautiTransportoPriemonesPavadinima() {
        return tps.gautiTransportoPriemonesPavadinima();
    }

    public TransportoPasirinkimoStrategija getTransportoPriemone() {
        return tps;
    }

    public String getAtvykimoTaskas() {
        return atvykimoTaskas;
    }

    public double getAtstumas() {
        return atstumas;
    }

    public Double getLaikas() {
        return laikas;
    }

    public BigDecimal getKaina() {
        return kaina;
    }

    public Intervalas getIlgalaikisIntervalas() {
        return ilgalaikisIntervalas;
    }
}